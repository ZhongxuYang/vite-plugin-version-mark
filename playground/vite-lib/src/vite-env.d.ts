/// <reference types="vite/client" />

declare const __LIB_VERSION__: string
