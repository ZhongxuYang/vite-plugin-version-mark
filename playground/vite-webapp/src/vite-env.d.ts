/// <reference types="vite/client" />

declare const __VITE_WEBAPP_VERSION__: string
