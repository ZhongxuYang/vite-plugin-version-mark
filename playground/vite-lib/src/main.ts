import './style.css'
import typescriptLogo from './typescript.svg'

document.querySelector<HTMLDivElement>('#app')!.innerHTML = `
  <div>
    <a href="https://vitejs.dev" target="_blank">
      <img src="/vite.svg" class="logo" alt="Vite logo" />
    </a>
    <a href="https://www.typescriptlang.org/" target="_blank">
      <img src="${typescriptLogo}" class="logo vanilla" alt="TypeScript logo" />
    </a>
    <h1>Vite + TypeScript</h1>
    <div class="card">
      <span>version: ${__LIB_VERSION__}</span>
    </div>
    <p class="read-the-docs">
      You can also type <i>__LIB_VERSION__</i> into the console to get the version number.
    </p>
  </div>
`
