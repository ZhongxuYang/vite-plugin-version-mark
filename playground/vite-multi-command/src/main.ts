// Multi-Command Version Mark Demo

// Display version information from global variables
function displayVersionInfo() {
  const examples = [
    {
      id: 'simple-version',
      globalVar: '__MULTI_COMMAND_SIMPLE_VERSION__',
      description: 'Simple multi-command with default separator',
    },
    {
      id: 'advanced-version', 
      globalVar: '__MULTI_COMMAND_ADVANCED_VERSION__',
      description: 'Advanced multi-command with custom format',
    },
    {
      id: 'error-demo-version',
      globalVar: '__MULTI_COMMAND_ERROR_DEMO_VERSION__',
      description: 'Error handling demonstration',
    },
    {
      id: 'separator-version',
      globalVar: '__MULTI_COMMAND_SEPARATOR_VERSION__',
      description: 'Custom separator example',
    },
  ]

  examples.forEach(example => {
    const element = document.getElementById(example.id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const version = (window as any)[example.globalVar]
    
    if (element) {
      if (version) {
        element.innerHTML = `
          <div class="success">‚úÖ ${version}</div>
          <small>Global variable: <code>window.${example.globalVar}</code></small>
        `
      } else {
        element.innerHTML = `
          <div class="error">‚ùå Version not found</div>
          <small>Expected global variable: <code>window.${example.globalVar}</code></small>
        `
      }
    }
  })
}

// Log all available version variables
function logVersionVariables() {
  console.group('üöÄ Multi-Command Version Information')
  
  const versionVars = Object.keys(window).filter(key => 
    key.includes('VERSION') && key.startsWith('__'),
  )
  
  if (versionVars.length > 0) {
    console.log('Available version variables:')
    versionVars.forEach(varName => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      console.log(`  ${varName}: ${(window as any)[varName]}`)
    })
  } else {
    console.warn('No version variables found. Make sure the plugin is configured correctly.')
  }
  
  console.groupEnd()
}

// Display meta tag information
function logMetaTags() {
  console.group('üìã Meta Tags Information')
  
  const metaTags = document.querySelectorAll('meta[name="application-name"]')
  if (metaTags.length > 0) {
    console.log('Found version meta tags:')
    metaTags.forEach((tag, index) => {
      console.log(`  Meta ${index + 1}: ${tag.getAttribute('content')}`)
    })
  } else {
    console.warn('No version meta tags found.')
  }
  
  console.groupEnd()
}

// Initialize demo
document.addEventListener('DOMContentLoaded', () => {
  console.log('üéâ Multi-Command Version Mark Demo Loaded!')
  
  // Display version information
  displayVersionInfo()
  
  // Log debug information
  logVersionVariables()
  logMetaTags()
  
  // Add some interactive features
  console.log('üí° Try these commands in the console:')
  console.log('  - Object.keys(window).filter(k => k.includes("VERSION"))')
  console.log('  - document.querySelectorAll(\'meta[name="application-name"]\')')
});

// Export for console access
// eslint-disable-next-line @typescript-eslint/no-explicit-any
(window as any).demoUtils = {
  displayVersionInfo,
  logVersionVariables,
  logMetaTags,
}
