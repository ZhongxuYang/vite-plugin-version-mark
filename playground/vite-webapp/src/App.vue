<script setup lang="ts">
import { computed } from 'vue'
import HelloWorld from './components/HelloWorld.vue'

const version = computed(()=>{
  return import.meta.env.__VITE_WEBAPP_VERSION_VERSION__
})
</script>

<template>
  <div>
    <a href="https://vitejs.dev" target="_blank">
      <img src="/vite.svg" class="logo" alt="Vite logo" />
    </a>
    <a href="https://vuejs.org/" target="_blank">
      <img src="./assets/vue.svg" class="logo vue" alt="Vue logo" />
    </a>
  </div>
  {{ version }}
  <HelloWorld msg="Vite + Vue" />
</template>

<style scoped>
.logo {
  height: 6em;
  padding: 1.5em;
  will-change: filter;
  transition: filter 300ms;
}
.logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}
.logo.vue:hover {
  filter: drop-shadow(0 0 2em #42b883aa);
}
</style>
