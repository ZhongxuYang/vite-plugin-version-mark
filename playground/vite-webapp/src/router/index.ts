import {createRouter, createWebHashHistory} from 'vue-router'
import type {RouteRecordRaw} from 'vue-router'

const routes: RouteRecordRaw[] = [
  {
    path: '/',
    name: 'HelloWorld',
    component: () => import('../components/HelloWorld.vue'),
  },
  {
    path: '/test',
    name: 'TestCom',
    component: () => import('../components/TestCom.vue'),
  },
]

const router = createRouter({
  history: createWebHashHistory('/'),
  routes,
})
export default router
