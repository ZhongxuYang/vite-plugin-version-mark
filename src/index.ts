export * from './plugins/vite'
export * from './plugins/nuxt3'