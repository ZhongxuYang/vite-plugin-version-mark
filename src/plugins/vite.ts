import {mkdir, writeFile} from 'node:fs/promises'
import {resolve,dirname} from 'node:path'
import {analyticOptions} from './core'
import type {VitePluginVersionMarkInput, VitePluginVersionMarkConfig} from './core'
import type {Plugin, IndexHtmlTransformResult} from 'vite'

export type {VitePluginVersionMarkInput}
export const vitePluginVersionMark: (options?: VitePluginVersionMarkInput) => Plugin = (options = {}) => {
  let versionMarkConfig: VitePluginVersionMarkConfig
  const getVersionMarkConfig = async () => {
    if (!versionMarkConfig) versionMarkConfig = await analyticOptions(options)
    return versionMarkConfig
  }
  let outDir: string

  return {
    name: 'vite-plugin-version-mark',

    async config() {
      const {
        ifGlobal,
        printName,
        printVersion,
      } = await getVersionMarkConfig()

      if (ifGlobal) {
        const keyName = `__${printName}__`
        return {
          define: {
            [keyName]: JSON.stringify(printVersion),
          },
        }
      }
    },

    async renderChunk(code, chunk) {
      if (chunk.isEntry) {
        const {
          ifExport,
          printName,
          printVersion,
        } = await getVersionMarkConfig()

        let modifiedCode = code
        if (ifExport) modifiedCode += `\nexport const ${printName} = '${printVersion}';`

        return {
          code: modifiedCode,
          map: null,
        }
      }
    },

    async transformIndexHtml() {
      const {
        ifLog,
        ifMeta,
        ifGlobal,
        printName,
        printVersion,
        printInfo,
      } = await getVersionMarkConfig()

      const els: IndexHtmlTransformResult = []
      if (ifMeta) {
        els.push({
          tag: 'meta',
          injectTo: 'head-prepend',
          attrs: {
            name: 'application-name',
            content: printInfo,
          },
        })
      }
      if (ifLog) {
        els.push({
          tag: 'script',
          injectTo: 'head',
          children: `console.log("${printInfo}")`,
        })
      }
      if (ifGlobal) {
        els.push({
          tag: 'script',
          injectTo: 'head',
          children: `__${printName}__ = "${printVersion}"`,
        })
      }

      return els
    },
    configResolved(config) {
      outDir = config.build.outDir
    },
    async closeBundle() {
      const {fileList} = await getVersionMarkConfig()
      if (!fileList.length) return
      await Promise.all(fileList.map(async ({path, content = ''}) => {
        try {
          const dir = dirname(path)
          await mkdir(resolve(outDir, dir), {recursive: true})
          const outputFilePath = resolve(outDir, path)
          await writeFile(outputFilePath, content)
          this.info(`Generate version file in ${outputFilePath}`)
        } catch (error) {
          this.error(`Failed to generate version file at ${path}: ${(error as Error).message}`)
        }
      }))
    },
  } as Plugin
}
