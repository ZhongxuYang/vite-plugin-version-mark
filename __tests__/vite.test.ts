import {resolve} from 'node:path'
import {describe, test, expect} from 'vitest'
import {build} from 'vite'
import type {RollupOutput, OutputAsset, OutputChunk} from 'rollup'
import {vitePluginVersionMark, type VitePluginVersionMarkInput} from '../src'

const entryPathForVite = './fixtures/vite'
const entryFilenameForVite = 'index.html'
const entryPathForLib = './fixtures/lib'
const entryFilenameForLib = 'index.js'

async function buildVite(pluginConfig: VitePluginVersionMarkInput, entryPath, entryFilename, buildOptions = {}) {
  const indexHtmlPath = resolve(__dirname, entryPath)

  const outputs = (await build({
    root: indexHtmlPath,
    plugins: [vitePluginVersionMark(pluginConfig)],
    ...buildOptions,
    // https://cn.vitejs.dev/config/shared-options.html#loglevel
    logLevel: 'error',
  })) as RollupOutput | RollupOutput[]

  const output = Array.isArray(outputs) ? outputs[0].output : outputs.output
  const file = output.find(
    item => item.fileName === entryFilename,
  ) as OutputAsset | OutputChunk

  const codeStr = file.source || file.code
  
  return await codeStr.toString()
}

// https://github.com/Applelo/unplugin-inject-preload/blob/main/test/vite.test.ts#L24
describe('VitePlugin', () => {
  describe('Name', () => {
    test('assigned', async () => {
      const output = await buildVite(
        {
          name: 'my-test-name',
          version: '1.0.0',
          ifMeta: true,
          ifLog: true,
          ifGlobal: true,
          ifExport: true,
        },
        entryPathForVite,
        entryFilenameForVite,
      )
      expect(output).toMatchSnapshot()
    })
  })

  describe('Assigned version', () => {
    test('output: None', async () => {
      const output = await buildVite(
        {
          version: '1.0.0',
          ifMeta: false,
          ifLog: false,
          ifGlobal: false,
          ifExport: false,
        },
        entryPathForVite,
        entryFilenameForVite,
      )
      expect(output).toMatchSnapshot()
    })
  
    test('output: Meta', async () => {
      const output = await buildVite(
        {
          version: '1.0.0',
          ifMeta: true,
          ifLog: false,
          ifGlobal: false,
          ifExport: false,
        },
        entryPathForVite,
        entryFilenameForVite,
      )
      expect(output).toMatchSnapshot()
    })
  
    test('output: Log', async () => {
      const output = await buildVite(
        {
          version: '1.0.0',
          ifMeta: false,
          ifLog: true,
          ifGlobal: false,
          ifExport: false,
        },
        entryPathForVite,
        entryFilenameForVite,
      )
      expect(output).toMatchSnapshot()
    })
  
    test('output: Global', async () => {
      const output = await buildVite(
        {
          version: '1.0.0',
          ifMeta: false,
          ifLog: false,
          ifGlobal: true,
          ifExport: false,
        },
        entryPathForVite,
        entryFilenameForVite,
      )
      expect(output).toMatchSnapshot()
    })
  
    test('output: Export', async () => {
      const output = await buildVite(
        {
          version: '1.0.0',
          ifMeta: false,
          ifLog: false,
          ifGlobal: false,
          ifExport: true,
        },
        entryPathForLib,
        entryFilenameForLib,
        {
          build: {
            lib: {
              entry: './index.ts',
              name: 'index',
              fileName: 'index',
              formats: ['es'],
            },
          },
        },
      )
      expect(output).toMatchSnapshot()
    })
  })

  // describe('longSHA', () => { })
  // describe('shortSHA', () => { })
  // describe('command', () => { })
  // describe('plugin usage', () => { })
})

// describe('NuxtPlugin', () => {})
