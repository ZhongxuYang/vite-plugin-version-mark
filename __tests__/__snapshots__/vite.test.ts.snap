// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VitePlugin > Assigned version > output: Export 1`] = `
"console.log("this is main.ts");
const o = "1.0.0";
export {
  o as VITE_PLUGIN_VERSION_MARK_VERSION
};
"
`;

exports[`VitePlugin > Assigned version > output: Global 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
    <script>__VITE_PLUGIN_VERSION_MARK_VERSION__ = "1.0.0"</script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`VitePlugin > Assigned version > output: Log 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
    <script>console.log("VITE_PLUGIN_VERSION_MARK_VERSION: 1.0.0")</script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`VitePlugin > Assigned version > output: Meta 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta name="application-name" content="VITE_PLUGIN_VERSION_MARK_VERSION: 1.0.0">

    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`VitePlugin > Assigned version > output: None 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`VitePlugin > Name > assigned 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta name="application-name" content="MY_TEST_NAME_VERSION: 1.0.0">

    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DvLNQ2tW.js"></script>
    <script>console.log("MY_TEST_NAME_VERSION: 1.0.0")</script>
    <script>__MY_TEST_NAME_VERSION__ = "1.0.0"</script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`vitePlugin > Assigned version & output > output: Export 1`] = `
"console.log("this is main.ts");
const o = "1.0.0";
export {
  o as VITE_PLUGIN_VERSION_MARK_VERSION
};
"
`;

exports[`vitePlugin > Assigned version & output > output: Global 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
    <script>__VITE_PLUGIN_VERSION_MARK_VERSION__ = "1.0.0"</script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`vitePlugin > Assigned version & output > output: Log 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
    <script>console.log("VITE_PLUGIN_VERSION_MARK_VERSION: 1.0.0")</script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`vitePlugin > Assigned version & output > output: Meta 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta name="application-name" content="VITE_PLUGIN_VERSION_MARK_VERSION: 1.0.0">

    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`vitePlugin > Assigned version & output > output: None 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`vitePlugin > name > name: Custom 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta name="application-name" content="MY_TEST_NAME_VERSION: 1.0.0">

    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DvLNQ2tW.js"></script>
    <script>console.log("MY_TEST_NAME_VERSION: 1.0.0")</script>
    <script>__MY_TEST_NAME_VERSION__ = "1.0.0"</script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`vitePlugin > version & output > output: Export 1`] = `
"console.log("this is main.ts");
const o = "1.0.0";
export {
  o as VITE_PLUGIN_VERSION_MARK_VERSION
};
"
`;

exports[`vitePlugin > version & output > output: Global 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
    <script>__VITE_PLUGIN_VERSION_MARK_VERSION__ = "1.0.0"</script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`vitePlugin > version & output > output: Log 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
    <script>console.log("VITE_PLUGIN_VERSION_MARK_VERSION: 1.0.0")</script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`vitePlugin > version & output > output: Meta 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta name="application-name" content="VITE_PLUGIN_VERSION_MARK_VERSION: 1.0.0">

    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;

exports[`vitePlugin > version & output > output: None 1`] = `
"<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + TS</title>
    <script type="module" crossorigin src="/assets/index-DUby6dX2.js"></script>
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
"
`;
